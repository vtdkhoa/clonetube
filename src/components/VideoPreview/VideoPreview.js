import React, { Component } from 'react'
import { Image } from 'semantic-ui-react'
import TimeAgo from 'javascript-time-ago'
import en from 'javascript-time-ago/locale/en'
import { Link } from 'react-router-dom'
import { formatNumber } from '../../services/number/number-format'
import { formatVideoDuration } from '../../services/time/time-format'
import PropTypes from 'prop-types'
import './VideoPreview.scss'

TimeAgo.locale(en)
const timeAgo = new TimeAgo('en-US')

class VideoPreview extends Component {
  render() {
    const { video } = this.props

    if (!video) {
      return <div/>
    }

    const horizontal = this.props.horizontal ? 'horizontal' : null
    const viewCountAndTimeString = VideoPreview.getFormattedViewAndTime(video)
    const duration = video.contentDetails ? video.contentDetails.duration : null
    const videoDuration = formatVideoDuration(duration)
    const expanded = this.props.expanded ? 'expanded' : null
    const description = this.props.expanded ? video.snippet.description : null

    return (
      <Link to={{ pathname: this.props.pathname, search: this.props.search }}>
        <div className={["video-preview", horizontal, expanded].join(' ')}>
          <div className="image-container">
            <Image src={video.snippet.thumbnails.medium.url}/>
            <div className="time-label">
              <span>{videoDuration}</span>
            </div>
          </div>
          <div className="video-info">
            <div className={["semi-bold", "show-max-two-lines", expanded].join(' ')}>
              {video.snippet.title}
            </div>
            <div className="video-preview-metadata-container">
              <div className="channel-title">
                {video.snippet.channelTitle}
              </div>
              <div className="view-count-and-time">
                {viewCountAndTimeString}
              </div>
              <div className="show-max-two-lines">
                {description}
              </div>
            </div>
          </div>
        </div>
      </Link>
    )
  }

  static getFormattedViewAndTime(video) {
    const publicationDate = new Date(video.snippet.publishedAt)
    const videoViewCount = video.statistics ? video.statistics.viewCount : null

    if (videoViewCount) {
      const videoViewCountShort = formatNumber(video.statistics.viewCount)
      return `${videoViewCountShort} views â€¢ ${timeAgo.format(publicationDate)}`
    }

    return ''
  }
}

VideoPreview.propTypes = {
  video: PropTypes.object,
  horizontal: PropTypes.bool,
  expanded: PropTypes.bool,
  pathname: PropTypes.string,
  search: PropTypes.string
}

export default VideoPreview