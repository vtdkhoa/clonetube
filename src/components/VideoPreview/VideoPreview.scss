@import '../../styles/shared.scss';

.video-preview {
  /* vertical */
  display: grid;
  grid: 118px auto / 210px;

  /* horizontal */
  &.horizontal {
    grid: auto / 210px auto;
    grid-column-gap: 4px;

    &.expanded {
      grid: auto / 246px auto;
    }

    .video-info {
      grid-row: 1 / 2;
      grid-column: 2 / 3;
    }
  }
}

.video-preview:hover {
  cursor: pointer;
  background-color: #ccc;
}

.image-container {
  position: relative;
  grid-row: 1 / 2;
  grid-column: 1 / 2;

  .time-label {
    position: absolute;
    background: $text-color-dark;
    bottom: 0;
    right: 0;
    opacity: 0.8;
    border-radius: 2px;
    font-weight: 500;
    color: white;
    margin: 4px;
    padding: 2px 4px;
    line-height: 12px;
  }
}

.video-info {
  grid-row: 2 / 3;
  grid-column: 1 / 2;
  color: $text-color-dark;
  padding: 2px;

  .video-preview-metadata-container {
    padding-top: 5px;
    font-size: 13px;
    color: #6e6e6e;
    line-height: 1.5rem;

    .channel-title {
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
  }

  .show-max-two-lines {
    overflow: hidden;
    line-height: 1.4em;
    max-height: 2.8em;
  }

  .semi-bold {
    &.expanded {
      font-weight: 400;
      font-size: 1.3rem;
    }

    font-weight: 600;
  }
}