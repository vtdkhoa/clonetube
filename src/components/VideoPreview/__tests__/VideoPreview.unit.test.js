import React from 'react'
import VideoPreview from '../VideoPreview'
import { shallow } from 'enzyme'

describe('VideoPreview', () => {
  test('renders', () => {
    const wrapper = shallow(<VideoPreview/>)
    expect(wrapper).toMatchSnapshot()
  })
})