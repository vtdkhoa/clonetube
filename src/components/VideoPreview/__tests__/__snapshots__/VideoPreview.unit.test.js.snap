// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPreview renders horizontally 1`] = `
<div
  className="video-preview horizontal"
>
  <div
    className="image-container"
  >
    <Image
      as="img"
      src="https://i.ytimg.com/vi/G0JKdFjWkLA/mqdefault.jpg"
      ui={true}
    />
    <div
      className="time-label"
    >
      <span />
    </div>
  </div>
  <div
    className="video-info"
  >
    <div
      className="semi-bold show-max-two-lines"
    >
      The Weeknd - Snowchild (Official Video)
    </div>
    <div
      className="video-preview-metadata-container"
    >
      <div
        className="channel-title"
      >
        TheWeekndVEVO
      </div>
      <div>
        <span />
      </div>
    </div>
  </div>
</div>
`;

exports[`VideoPreview renders vertically 1`] = `
<div
  className="video-preview "
>
  <div
    className="image-container"
  >
    <Image
      as="img"
      src="https://i.ytimg.com/vi/G0JKdFjWkLA/mqdefault.jpg"
      ui={true}
    />
    <div
      className="time-label"
    >
      <span />
    </div>
  </div>
  <div
    className="video-info"
  >
    <div
      className="semi-bold show-max-two-lines"
    >
      The Weeknd - Snowchild (Official Video)
    </div>
    <div
      className="video-preview-metadata-container"
    >
      <div
        className="channel-title"
      >
        TheWeekndVEVO
      </div>
      <div>
        <span />
      </div>
    </div>
  </div>
</div>
`;

exports[`VideoPreview renders with props video = {} 1`] = `
<div
  className="video-preview "
>
  <div
    className="image-container"
  >
    <Image
      as="img"
      src="https://i.ytimg.com/vi/G0JKdFjWkLA/mqdefault.jpg"
      ui={true}
    />
    <div
      className="time-label"
    >
      <span />
    </div>
  </div>
  <div
    className="video-info"
  >
    <div
      className="semi-bold show-max-two-lines"
    >
      The Weeknd - Snowchild (Official Video)
    </div>
    <div
      className="video-preview-metadata-container"
    >
      <div
        className="channel-title"
      >
        TheWeekndVEVO
      </div>
      <div>
        <span />
      </div>
    </div>
  </div>
</div>
`;
